core/impl/src/main/java/com/blazebit/persistence/impl/function/window/AbstractWindowFunction.java

```diff
@@ -0,0 +1,557 @@
+/*
+ * Copyright 2014 - 2019 Blazebit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.blazebit.persistence.impl.function.window;
+
+import com.blazebit.persistence.spi.FunctionRenderContext;
+import com.blazebit.persistence.spi.JpqlFunction;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ *
+ * @author Jan-Willem Gmelig Meyling
+ * @author Sayra Ranjha
+ * @since 1.4.0
+ */
+public abstract class AbstractWindowFunction implements JpqlFunction {
+
```

Comment:
IMO a constructor should capture the window support level which is then provided on a per DBMS level. Also the function name should IMO be passed as constructor argument.
Something like this

```
publict AbstractWindowFunction(String functionName, DbmsDialect dbmsDialect) {
  //...
}
```

And then use newly introduced methods e.g. `dbmsDialect.supportsFilter()`, `dbmsDialect.supportsNullPrecedence()`

Reply:
Agreed. I actually was suspicious why we didnt already do that. Group concat could benefit from this as well. However, the actual dialect used (with any potential configuration) is not available at construction time. Not sure how many dialects actually use the extra constructor with additional configuration/types.

Reply:
I have addressed this comment.